File: FW/lunarcore/src/onion.rs
use crate::crypto::{
    x25519::x25519,
    hkdf::Hkdf,
    aes::Aes256,
    sha256::Sha256,
};
use crate::transport::{NODE_HINT_SIZE, AUTH_TAG_SIZE};
use heapless::Vec as HeaplessVec;

#[inline(never)]
fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }

    let mut result: u8 = 0;
    for (x, y) in a.iter().zip(b.iter()) {
        result |= x ^ y;
    }

    unsafe {
        core::ptr::read_volatile(&result) == 0
    }
}

pub const MAX_HOPS: usize = 7;

pub const MIN_HOPS: usize = 3;

pub const HOP_OVERHEAD: usize = NODE_HINT_SIZE + AUTH_TAG_SIZE;

const ONION_KEY_INFO: &[u8] = b"lunarpunk-onion-key-v1";

const ONION_BLIND_INFO: &[u8] = b"lunarpunk-onion-blind-v1";

#[derive(Debug, Clone)]
pub struct RouteHop {

    pub hint: u16,

    pub public_key: [u8; 32],
}

#[derive(Debug, Clone)]
pub struct OnionRoute {

    pub hops: HeaplessVec<RouteHop, MAX_HOPS>,
}

impl OnionRoute {

    pub fn new(hops: &[RouteHop]) -> Option<Self> {
        if hops.len() < MIN_HOPS || hops.len() > MAX_HOPS {
            return None;
        }

        let mut route_hops = HeaplessVec::new();
        for hop in hops {
            route_hops.push(hop.clone()).ok()?;
        }

        Some(Self { hops: route_hops })
    }

    pub fn len(&self) -> usize {
        self.hops.len()
    }

    pub fn is_empty(&self) -> bool {
        self.hops.is_empty()
    }

    pub fn entry_hint(&self) -> u16 {
        self.hops.first().map(|h| h.hint).unwrap_or(0)
    }

    pub fn exit_hint(&self) -> u16 {
        self.hops.last().map(|h| h.hint).unwrap_or(0)
    }

    pub fn overhead(&self) -> usize {
        self.hops.len() * HOP_OVERHEAD
    }
}

#[derive(Debug, Clone)]
pub struct OnionPacket {

    pub data: HeaplessVec<u8, 256>,

    pub num_layers: u8,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OnionError {

    InvalidRoute,

    PacketTooLarge,

    AuthenticationFailed,

    DecryptionFailed,

    NoMoreLayers,
}

pub struct OnionRouter {

    our_private: [u8; 32],

    our_public: [u8; 32],

    our_hint: u16,
}

impl OnionRouter {

    pub fn new(private_key: [u8; 32]) -> Self {
        use crate::crypto::x25519::x25519_base;

        let our_public = x25519_base(&private_key);

        let hint_hash = Sha256::hash(&our_public);
        let our_hint = ((hint_hash[0] as u16) << 8) | (hint_hash[1] as u16);

        Self {
            our_private: private_key,
            our_public,
            our_hint,
        }
    }

    pub fn wrap(&self, payload: &[u8], route: &OnionRoute) -> Result<OnionPacket, OnionError> {
        if route.len() < MIN_HOPS || route.len() > MAX_HOPS {
            return Err(OnionError::InvalidRoute);
        }

        let total_overhead = route.overhead();
        if payload.len() + total_overhead > 256 {
            return Err(OnionError::PacketTooLarge);
        }

        let mut current = HeaplessVec::<u8, 256>::new();
        current.extend_from_slice(payload).map_err(|_| OnionError::PacketTooLarge)?;

        for (i, hop) in route.hops.iter().enumerate().rev() {

            let shared = x25519(&self.our_private, &hop.public_key);

            let layer_index = (route.len() - 1 - i) as u8;
            let mut key_input = [0u8; 33];
            key_input[..32].copy_from_slice(&shared);
            key_input[32] = layer_index;

            let mut layer_key = [0u8; 32];
            Hkdf::derive(&key_input, &[layer_index], ONION_KEY_INFO, &mut layer_key);

            let encrypted = self.encrypt_layer(&layer_key, &current)?;

            let tag = self.compute_tag(&layer_key, &encrypted);

            let mut new_layer = HeaplessVec::<u8, 256>::new();

            let next_hint = if i == route.len() - 1 {

                0u16
            } else {
                route.hops[i + 1].hint
            };

            new_layer.push((next_hint >> 8) as u8).map_err(|_| OnionError::PacketTooLarge)?;
            new_layer.push(next_hint as u8).map_err(|_| OnionError::PacketTooLarge)?;
            new_layer.extend_from_slice(&tag).map_err(|_| OnionError::PacketTooLarge)?;
            new_layer.extend_from_slice(&encrypted).map_err(|_| OnionError::PacketTooLarge)?;

            current = new_layer;
        }

        Ok(OnionPacket {
            data: current,
            num_layers: route.len() as u8,
        })
    }

    pub fn unwrap(&self, packet: &OnionPacket, sender_public: &[u8; 32]) -> Result<(u16, OnionPacket), OnionError> {
        if packet.data.len() < HOP_OVERHEAD {
            return Err(OnionError::DecryptionFailed);
        }

        let next_hint = ((packet.data[0] as u16) << 8) | (packet.data[1] as u16);
        let tag = &packet.data[2..18];
        let encrypted = &packet.data[18..];

        let shared = x25519(&self.our_private, sender_public);

        let layer_index = packet.num_layers.saturating_sub(1);
        let mut key_input = [0u8; 33];
        key_input[..32].copy_from_slice(&shared);
        key_input[32] = layer_index;

        let mut layer_key = [0u8; 32];
        Hkdf::derive(&key_input, &[layer_index], ONION_KEY_INFO, &mut layer_key);

        let expected_tag = self.compute_tag(&layer_key, encrypted);
        if !constant_time_eq(tag, &expected_tag) {
            return Err(OnionError::AuthenticationFailed);
        }

        let inner = self.decrypt_layer(&layer_key, encrypted)?;

        if next_hint == 0 {
            return Err(OnionError::NoMoreLayers);
        }

        let mut inner_packet = OnionPacket {
            data: inner,
            num_layers: packet.num_layers.saturating_sub(1),
        };

        Ok((next_hint, inner_packet))
    }

    pub fn unwrap_final(&self, packet: &OnionPacket, sender_public: &[u8; 32]) -> Result<HeaplessVec<u8, 256>, OnionError> {
        if packet.data.len() < HOP_OVERHEAD {
            return Err(OnionError::DecryptionFailed);
        }

        let tag = &packet.data[2..18];
        let encrypted = &packet.data[18..];

        let shared = x25519(&self.our_private, sender_public);

        let layer_index = packet.num_layers.saturating_sub(1);
        let mut key_input = [0u8; 33];
        key_input[..32].copy_from_slice(&shared);
        key_input[32] = layer_index;

        let mut layer_key = [0u8; 32];
        Hkdf::derive(&key_input, &[layer_index], ONION_KEY_INFO, &mut layer_key);

        let expected_tag = self.compute_tag(&layer_key, encrypted);
        if !constant_time_eq(tag, &expected_tag) {
            return Err(OnionError::AuthenticationFailed);
        }

        self.decrypt_layer(&layer_key, encrypted)
    }

    fn encrypt_layer(&self, key: &[u8; 32], data: &[u8]) -> Result<HeaplessVec<u8, 256>, OnionError> {
        let mut output = HeaplessVec::new();
        output.extend_from_slice(data).map_err(|_| OnionError::PacketTooLarge)?;

        let aes = Aes256::new(key);
        let mut counter = [0u8; 16];
        let mut keystream = [0u8; 16];
        let mut block_num = 0u64;

        for chunk in output.chunks_mut(16) {
            keystream.copy_from_slice(&counter);
            keystream[8..].copy_from_slice(&block_num.to_le_bytes());
            aes.encrypt_block(&mut keystream);
            for (c, k) in chunk.iter_mut().zip(keystream.iter()) {
                *c ^= k;
            }
            block_num += 1;
        }

        Ok(output)
    }

    fn decrypt_layer(&self, key: &[u8; 32], data: &[u8]) -> Result<HeaplessVec<u8, 256>, OnionError> {

        self.encrypt_layer(key, data)
    }

    fn compute_tag(&self, key: &[u8; 32], data: &[u8]) -> [u8; 16] {

        let mut inner = [0x36u8; 64];
        let mut outer = [0x5cu8; 64];

        for (i, k) in key.iter().enumerate() {
            inner[i] ^= k;
            outer[i] ^= k;
        }

        let mut inner_input = HeaplessVec::<u8, 320>::new();
        let _ = inner_input.extend_from_slice(&inner);
        let _ = inner_input.extend_from_slice(data);
        let inner_hash = Sha256::hash(&inner_input);

        let mut outer_input = [0u8; 96];
        outer_input[..64].copy_from_slice(&outer);
        outer_input[64..].copy_from_slice(&inner_hash);
        let full = Sha256::hash(&outer_input);

        let mut tag = [0u8; 16];
        tag.copy_from_slice(&full[..16]);
        tag
    }

    pub fn our_hint(&self) -> u16 {
        self.our_hint
    }

    pub fn our_public(&self) -> &[u8; 32] {
        &self.our_public
    }

    pub fn derive_blinded_hint(&self, epoch: u64) -> u16 {
        let mut input = [0u8; 40];
        input[..32].copy_from_slice(&self.our_public);
        input[32..].copy_from_slice(&epoch.to_le_bytes());

        let mut blinded = [0u8; 2];
        Hkdf::derive(&input, &epoch.to_le_bytes(), ONION_BLIND_INFO, &mut blinded);

        ((blinded[0] as u16) << 8) | (blinded[1] as u16)
    }
}

pub struct RouteBuilder {

    relays: HeaplessVec<RouteHop, 32>,
}

impl RouteBuilder {
    pub fn new() -> Self {
        Self {
            relays: HeaplessVec::new(),
        }
    }

    pub fn add_relay(&mut self, hint: u16, public_key: [u8; 32]) -> bool {
        self.relays.push(RouteHop { hint, public_key }).is_ok()
    }

    pub fn build_route(&self, destination: RouteHop, num_hops: usize) -> Option<OnionRoute> {
        if num_hops < MIN_HOPS || num_hops > MAX_HOPS {
            return None;
        }

        let relay_count = num_hops - 1;
        if self.relays.len() < relay_count {
            return None;
        }

        let mut hops = HeaplessVec::<RouteHop, MAX_HOPS>::new();

        for i in 0..relay_count {
            hops.push(self.relays[i % self.relays.len()].clone()).ok()?;
        }

        hops.push(destination).ok()?;

        Some(OnionRoute { hops })
    }

    pub fn relay_count(&self) -> usize {
        self.relays.len()
    }
}

impl Default for RouteBuilder {
    fn default() -> Self {
        Self::new()
    }
}

File: FW/lunarcore_c/onion.c
#include "onion.h"

// Dependencies from crypto module
extern void x25519(const uint8_t private_key[32], const uint8_t public_key[32], uint8_t shared[32]);
extern void x25519_base(const uint8_t private_key[32], uint8_t public_key[32]);
extern void hkdf_derive(const uint8_t *input_key, size_t input_key_len,
                        const uint8_t *salt, size_t salt_len,
                        const uint8_t *info, size_t info_len,
                        uint8_t *output, size_t output_len);
extern void sha256_hash(const uint8_t *data, size_t len, uint8_t output[32]);
extern void aes256_new(const uint8_t key[32], void **aes_ctx);
extern void aes256_encrypt_block(void *aes_ctx, uint8_t block[16]);
extern void aes256_free(void *aes_ctx);

// Dependencies from transport module
#define NODE_HINT_SIZE 2
#define AUTH_TAG_SIZE 16

// HeaplessVec implementation for uint8_t with max capacity 256
typedef struct 
{
  uint8_t data[256];
  size_t len;
  size_t capacity;
} HeaplessVec_u8_256;

static inline bool heapless_vec_u8_256_init(HeaplessVec_u8_256 *vec) 
{
  vec->len = 0;
  vec->capacity = 256;
  memset(vec->data, 0, 256);
  return true;
}

static inline bool heapless_vec_u8_256_push(HeaplessVec_u8_256 *vec, uint8_t val) 
{
  if (vec->len >= vec->capacity) return false;
  vec->data[vec->len] = val;
  vec->len++;
  return true;
}

static inline bool heapless_vec_u8_256_extend_from_slice(HeaplessVec_u8_256 *vec, const uint8_t *data, size_t len) 
{
  if (vec->len + len > vec->capacity) return false;
  memcpy(&vec->data[vec->len], data, len);
  vec->len += len;
  return true;
}

static inline uint8_t* heapless_vec_u8_256_as_slice(HeaplessVec_u8_256 *vec) 
{
  return vec->data;
}

static inline size_t heapless_vec_u8_256_len(const HeaplessVec_u8_256 *vec) 
{
  return vec->len;
}

// HeaplessVec implementation for uint8_t with max capacity 320
typedef struct 
{
  uint8_t data[320];
  size_t len;
  size_t capacity;
} HeaplessVec_u8_320;

static inline bool heapless_vec_u8_320_init(HeaplessVec_u8_320 *vec) 
{
  vec->len = 0;
  vec->capacity = 320;
  memset(vec->data, 0, 320);
  return true;
}

static inline bool heapless_vec_u8_320_extend_from_slice(HeaplessVec_u8_320 *vec, const uint8_t *data, size_t len) 
{
  if (vec->len + len > vec->capacity) return false;
  memcpy(&vec->data[vec->len], data, len);
  vec->len += len;
  return true;
}

// Constant time equality check
static inline bool constant_time_eq(const uint8_t *a, size_t a_len, const uint8_t *b, size_t b_len) 
{
  if (a_len != b_len) return false;
  uint8_t result = 0;
  for (size_t i = 0; i < a_len; i++) result |= a[i] ^ b[i];
  // Force volatile read to prevent optimization
  volatile uint8_t volatile_result = result;
  return volatile_result == 0;
}

// Constants
#define MAX_HOPS 7
#define MIN_HOPS 3
#define HOP_OVERHEAD (NODE_HINT_SIZE + AUTH_TAG_SIZE)

static const uint8_t ONION_KEY_INFO[] = "lunarpunk-onion-key-v1";
static const size_t ONION_KEY_INFO_LEN = 22;

static const uint8_t ONION_BLIND_INFO[] = "lunarpunk-onion-blind-v1";
static const size_t ONION_BLIND_INFO_LEN = 24;

// RouteHop structure
typedef struct 
{
  uint16_t hint;
  uint8_t public_key[32];
} RouteHop;

// HeaplessVec for RouteHop with max capacity MAX_HOPS
typedef struct 
{
  RouteHop data[MAX_HOPS];
  size_t len;
  size_t capacity;
} HeaplessVec_RouteHop_MAX_HOPS;

static inline bool heapless_vec_routehop_init(HeaplessVec_RouteHop_MAX_HOPS *vec) 
{
  vec->len = 0;
  vec->capacity = MAX_HOPS;
  memset(vec->data, 0, sizeof(RouteHop) * MAX_HOPS);
  return true;
}

static inline bool heapless_vec_routehop_push(HeaplessVec_RouteHop_MAX_HOPS *vec, const RouteHop *hop) 
{
  if (vec->len >= vec->capacity) return false;
  memcpy(&vec->data[vec->len], hop, sizeof(RouteHop));
  vec->len++;
  return true;
}

static inline const RouteHop* heapless_vec_routehop_first(const HeaplessVec_RouteHop_MAX_HOPS *vec) 
{
  if (vec->len == 0) return NULL;
  return &vec->data[0];
}

static inline const RouteHop* heapless_vec_routehop_last(const HeaplessVec_RouteHop_MAX_HOPS *vec) 
{
  if (vec->len == 0) return NULL;
  return &vec->data[vec->len - 1];
}

// OnionRoute structure
typedef struct 
{
  HeaplessVec_RouteHop_MAX_HOPS hops;
} OnionRoute;

// OnionRoute functions
static inline bool onion_route_new(OnionRoute *route, const RouteHop *hops, size_t hops_len) 
{
  if (hops_len < MIN_HOPS || hops_len > MAX_HOPS) return false;
  heapless_vec_routehop_init(&route->hops);
  for (size_t i = 0; i < hops_len; i++) 
  {
    if (!heapless_vec_routehop_push(&route->hops, &hops[i])) return false;
  }
  return true;
}

static inline size_t onion_route_len(const OnionRoute *route) 
{
  return route->hops.len;
}

static inline bool onion_route_is_empty(const OnionRoute *route) 
{
  return route->hops.len == 0;
}

static inline uint16_t onion_route_entry_hint(const OnionRoute *route) 
{
  const RouteHop *first = heapless_vec_routehop_first(&route->hops);
  if (first == NULL) return 0;
  return first->hint;
}

static inline uint16_t onion_route_exit_hint(const OnionRoute *route) 
{
  const RouteHop *last = heapless_vec_routehop_last(&route->hops);
  if (last == NULL) return 0;
  return last->hint;
}

static inline size_t onion_route_overhead(const OnionRoute *route) 
{
  return route->hops.len * HOP_OVERHEAD;
}

// OnionPacket structure
typedef struct 
{
  HeaplessVec_u8_256 data;
  uint8_t num_layers;
} OnionPacket;

// OnionError enumeration
typedef enum 
{
  ONION_ERROR_INVALID_ROUTE,
  ONION_ERROR_PACKET_TOO_LARGE,
  ONION_ERROR_AUTHENTICATION_FAILED,
  ONION_ERROR_DECRYPTION_FAILED,
  ONION_ERROR_NO_MORE_LAYERS
} OnionError;

// OnionRouter structure
typedef struct 
{
  uint8_t our_private[32];
  uint8_t our_public[32];
  uint16_t our_hint;
} OnionRouter;

// OnionRouter constructor
static inline void onion_router_new(OnionRouter *router, const uint8_t private_key[32]) 
{
  memcpy(router->our_private, private_key, 32);
  x25519_base(private_key, router->our_public);
  uint8_t hint_hash[32];
  sha256_hash(router->our_public, 32, hint_hash);
  router->our_hint = ((uint16_t)hint_hash[0] << 8) | (uint16_t)hint_hash[1];
}

// Forward declarations for OnionRouter helper functions
static bool onion_router_encrypt_layer(OnionRouter *self, const uint8_t key[32], 
                                       const uint8_t *data, size_t data_len,
                                       HeaplessVec_u8_256 *output);
static bool onion_router_decrypt_layer(OnionRouter *self, const uint8_t key[32],
                                       const uint8_t *data, size_t data_len,
                                       HeaplessVec_u8_256 *output);
static void onion_router_compute_tag(OnionRouter *self, const uint8_t key[32],
                                     const uint8_t *data, size_t data_len,
                                     uint8_t tag[16]);

// OnionRouter wrap function
static inline bool onion_router_wrap(OnionRouter *self, const uint8_t *payload, size_t payload_len,
                                     const OnionRoute *route, OnionPacket *packet,
                                     OnionError *error) 
{
  if (route->hops.len < MIN_HOPS || route->hops.len > MAX_HOPS) 
  {
    if (error) *error = ONION_ERROR_INVALID_ROUTE;
    return false;
  }
  size_t total_overhead = onion_route_overhead(route);
  if (payload_len + total_overhead > 256) 
  {
    if (error) *error = ONION_ERROR_PACKET_TOO_LARGE;
    return false;
  }
  HeaplessVec_u8_256 current;
  heapless_vec_u8_256_init(&current);
  if (!heapless_vec_u8_256_extend_from_slice(&current, payload, payload_len)) 
  {
    if (error) *error = ONION_ERROR_PACKET_TOO_LARGE;
    return false;
  }
  // Process hops in reverse order
  for (int i = (int)route->hops.len - 1; i >= 0; i--) 
  {
    const RouteHop *hop = &route->hops.data[i];
    // Compute shared secret
    uint8_t shared[32];
    x25519(self->our_private, hop->public_key, shared);
    // Compute layer index
    uint8_t layer_index = (uint8_t)(route->hops.len - 1 - i);
    uint8_t key_input[33];
    memcpy(key_input, shared, 32);
    key_input[32] = layer_index;
    // Derive layer key
    uint8_t layer_key[32];
    uint8_t salt[1] = { layer_index };
    hkdf_derive(key_input, 33, salt, 1, ONION_KEY_INFO, ONION_KEY_INFO_LEN, layer_key, 32);
    // Encrypt current layer
    HeaplessVec_u8_256 encrypted;
    heapless_vec_u8_256_init(&encrypted);
    if (!onion_router_encrypt_layer(self, layer_key, current.data, current.len, &encrypted)) 
    {
      if (error) *error = ONION_ERROR_PACKET_TOO_LARGE;
      return false;
    }
    // Compute authentication tag
    uint8_t tag[16];
    onion_router_compute_tag(self, layer_key, encrypted.data, encrypted.len, tag);
    // Build new layer
    HeaplessVec_u8_256 new_layer;
    heapless_vec_u8_256_init(&new_layer);
    // Determine next hint
    uint16_t next_hint;
    if (i == (int)route->hops.len - 1) next_hint = 0;
    else next_hint = route->hops.data[i + 1].hint;
    // Add next hint (2 bytes)
    if (!heapless_vec_u8_256_push(&new_layer, (uint8_t)(next_hint >> 8))) 
    {
      if (error) *error = ONION_ERROR_PACKET_TOO_LARGE;
      return false;
    }
    if (!heapless_vec_u8_256_push(&new_layer, (uint8_t)next_hint)) 
    {
      if (error) *error = ONION_ERROR_PACKET_TOO_LARGE;
      return false;
    }
    // Add authentication tag
    if (!heapless_vec_u8_256_extend_from_slice(&new_layer, tag, 16)) 
    {
      if (error) *error = ONION_ERROR_PACKET_TOO_LARGE;
      return false;
    }
    // Add encrypted data
    if (!heapless_vec_u8_256_extend_from_slice(&new_layer, encrypted.data, encrypted.len)) 
    {
      if (error) *error = ONION_ERROR_PACKET_TOO_LARGE;
      return false;
    }
    current = new_layer;
  }
  packet->data = current;
  packet->num_layers = (uint8_t)route->hops.len;
  return true;
}

// OnionRouter unwrap function
static inline bool onion_router_unwrap(OnionRouter *self, const OnionPacket *packet,
                                      const uint8_t sender_public[32],
                                      uint16_t *next_hint_out, OnionPacket *inner_packet_out,
                                      OnionError *error) 
{
  if (packet->data.len < HOP_OVERHEAD) 
  {
    if (error) *error = ONION_ERROR_DECRYPTION_FAILED;
    return false;
  }
  uint16_t next_hint = ((uint16_t)packet->data.data[0] << 8) | (uint16_t)packet->data.data[1];
  const uint8_t *tag = &packet->data.data[2];
  const uint8_t *encrypted = &packet->data.data[18];
  size_t encrypted_len = packet->data.len - 18;
  // Compute shared secret
  uint8_t shared[32];
  x25519(self->our_private, sender_public, shared);
  // Compute layer index
  uint8_t layer_index = packet->num_layers > 0 ? packet->num_layers - 1 : 0;
  uint8_t key_input[33];
  memcpy(key_input, shared, 32);
  key_input[32] = layer_index;
  // Derive layer key
  uint8_t layer_key[32];
  uint8_t salt[1] = { layer_index };
  hkdf_derive(key_input, 33, salt, 1, ONION_KEY_INFO, ONION_KEY_INFO_LEN, layer_key, 32);
  // Verify authentication tag
  uint8_t expected_tag[16];
  onion_router_compute_tag(self, layer_key, encrypted, encrypted_len, expected_tag);
  if (!constant_time_eq(tag, 16, expected_tag, 16)) 
  {
    if (error) *error = ONION_ERROR_AUTHENTICATION_FAILED;
    return false;
  }
  // Decrypt layer
  HeaplessVec_u8_256 inner;
  heapless_vec_u8_256_init(&inner);
  if (!onion_router_decrypt_layer(self, layer_key, encrypted, encrypted_len, &inner)) 
  {
    if (error) *error = ONION_ERROR_DECRYPTION_FAILED;
    return false;
  }
  // Check if this is the final layer
  if (next_hint == 0) 
  {
    if (error) *error = ONION_ERROR_NO_MORE_LAYERS;
    return false;
  }
  // Return results
  *next_hint_out = next_hint;
  inner_packet_out->data = inner;
  inner_packet_out->num_layers = packet->num_layers > 0 ? packet->num_layers - 1 : 0;
  return true;
}

// OnionRouter unwrap_final function
static inline bool onion_router_unwrap_final(OnionRouter *self, const OnionPacket *packet,
                                             const uint8_t sender_public[32],
                                             HeaplessVec_u8_256 *payload_out,
                                             OnionError *error) 
{
  if (packet->data.len < HOP_OVERHEAD) 
  {
    if (error) *error = ONION_ERROR_DECRYPTION_FAILED;
    return false;
  }
  const uint8_t *tag = &packet->data.data[2];
  const uint8_t *encrypted = &packet->data.data[18];
  size_t encrypted_len = packet->data.len - 18;
  // Compute shared secret
  uint8_t shared[32];
  x25519(self->our_private, sender_public, shared);
  // Compute layer index
  uint8_t layer_index = packet->num_layers > 0 ? packet->num_layers - 1 : 0;
  uint8_t key_input[33];
  memcpy(key_input, shared, 32);
  key_input[32] = layer_index;
  // Derive layer key
  uint8_t layer_key[32];
  uint8_t salt[1] = { layer_index };
  hkdf_derive(key_input, 33, salt, 1, ONION_KEY_INFO, ONION_KEY_INFO_LEN, layer_key, 32);
  // Verify authentication tag
  uint8_t expected_tag[16];
  onion_router_compute_tag(self, layer_key, encrypted, encrypted_len, expected_tag);
  if (!constant_time_eq(tag, 16, expected_tag, 16)) 
  {
    if (error) *error = ONION_ERROR_AUTHENTICATION_FAILED;
    return false;
  }
  // Decrypt layer
  heapless_vec_u8_256_init(payload_out);
  if (!onion_router_decrypt_layer(self, layer_key, encrypted, encrypted_len, payload_out)) 
  {
    if (error) *error = ONION_ERROR_DECRYPTION_FAILED;
    return false;
  }
  return true;
}

// OnionRouter encrypt_layer implementation
static bool onion_router_encrypt_layer(OnionRouter *self, const uint8_t key[32],
                                       const uint8_t *data, size_t data_len,
                                       HeaplessVec_u8_256 *output) 
{
  heapless_vec_u8_256_init(output);
  if (!heapless_vec_u8_256_extend_from_slice(output, data, data_len)) return false;
  void *aes_ctx;
  aes256_new(key, &aes_ctx);
  uint8_t counter[16];
  memset(counter, 0, 16);
  uint8_t keystream[16];
  uint64_t block_num = 0;
  // Process data in 16-byte chunks
  for (size_t offset = 0; offset < output->len; offset += 16) 
  {
    size_t chunk_size = (output->len - offset < 16) ? (output->len - offset) : 16;
    // Prepare keystream block
    memcpy(keystream, counter, 16);
    memcpy(&keystream[8], &block_num, 8);
    // Encrypt the counter to generate keystream
    aes256_encrypt_block(aes_ctx, keystream);
    // XOR data with keystream
    for (size_t i = 0; i < chunk_size; i++) output->data[offset + i] ^= keystream[i];
    block_num++;
  }
  aes256_free(aes_ctx);
  return true;
}

// OnionRouter decrypt_layer implementation (same as encrypt due to CTR mode)
static bool onion_router_decrypt_layer(OnionRouter *self, const uint8_t key[32],
                                       const uint8_t *data, size_t data_len,
                                       HeaplessVec_u8_256 *output) 
{
  // CTR mode encryption and decryption are the same operation
  return onion_router_encrypt_layer(self, key, data, data_len, output);
}

// OnionRouter compute_tag implementation (HMAC-SHA256 truncated to 16 bytes)
static void onion_router_compute_tag(OnionRouter *self, const uint8_t key[32],
                                     const uint8_t *data, size_t data_len,
                                     uint8_t tag[16]) 
{
  // HMAC-SHA256 implementation
  uint8_t inner[64];
  uint8_t outer[64];
  // Initialize with padding
  memset(inner, 0x36, 64);
  memset(outer, 0x5c, 64);
  // XOR key with padding
  for (size_t i = 0; i < 32; i++) 
  {
    inner[i] ^= key[i];
    outer[i] ^= key[i];
  }
  // Inner hash: H(K XOR ipad || data)
  HeaplessVec_u8_320 inner_input;
  heapless_vec_u8_320_init(&inner_input);
  heapless_vec_u8_320_extend_from_slice(&inner_input, inner, 64);
  heapless_vec_u8_320_extend_from_slice(&inner_input, data, data_len);
  uint8_t inner_hash[32];
  sha256_hash(inner_input.data, inner_input.len, inner_hash);
  // Outer hash: H(K XOR opad || inner_hash)
  uint8_t outer_input[96];
  memcpy(outer_input, outer, 64);
  memcpy(&outer_input[64], inner_hash, 32);
  uint8_t full[32];
  sha256_hash(outer_input, 96, full);
  // Truncate to 16 bytes
  memcpy(tag, full, 16);
}

// OnionRouter getter functions
static inline uint16_t onion_router_our_hint(const OnionRouter *router) 
{
  return router->our_hint;
}

static inline const uint8_t* onion_router_our_public(const OnionRouter *router) 
{
  return router->our_public;
}

// OnionRouter derive_blinded_hint function
static inline uint16_t onion_router_derive_blinded_hint(const OnionRouter *router, uint64_t epoch) 
{
  uint8_t input[40];
  memcpy(input, router->our_public, 32);
  memcpy(&input[32], &epoch, 8);
  uint8_t blinded[2];
  uint8_t epoch_bytes[8];
  memcpy(epoch_bytes, &epoch, 8);
  hkdf_derive(input, 40, epoch_bytes, 8, ONION_BLIND_INFO, ONION_BLIND_INFO_LEN, blinded, 2);
  return ((uint16_t)blinded[0] << 8) | (uint16_t)blinded[1];
}

// HeaplessVec for RouteHop with max capacity 32
typedef struct 
{
  RouteHop data[32];
  size_t len;
  size_t capacity;
} HeaplessVec_RouteHop_32;

static inline bool heapless_vec_routehop_32_init(HeaplessVec_RouteHop_32 *vec) 
{
  vec->len = 0;
  vec->capacity = 32;
  memset(vec->data, 0, sizeof(RouteHop) * 32);
  return true;
}

static inline bool heapless_vec_routehop_32_push(HeaplessVec_RouteHop_32 *vec, const RouteHop *hop) 
{
  if (vec->len >= vec->capacity) return false;
  memcpy(&vec->data[vec->len], hop, sizeof(RouteHop));
  vec->len++;
  return true;
}

// RouteBuilder structure
typedef struct 
{
  HeaplessVec_RouteHop_32 relays;
} RouteBuilder;

// RouteBuilder constructor
static inline void route_builder_new(RouteBuilder *builder) 
{
  heapless_vec_routehop_32_init(&builder->relays);
}

// RouteBuilder add_relay function
static inline bool route_builder_add_relay(RouteBuilder *builder, uint16_t hint, const uint8_t public_key[32]) 
{
  RouteHop hop;
  hop.hint = hint;
  memcpy(hop.public_key, public_key, 32);
  return heapless_vec_routehop_32_push(&builder->relays, &hop);
}

// RouteBuilder build_route function
static inline bool route_builder_build_route(const RouteBuilder *builder, const RouteHop *destination,
                                             size_t num_hops, OnionRoute *route_out) 
{
  if (num_hops < MIN_HOPS || num_hops > MAX_HOPS) return false;
  size_t relay_count = num_hops - 1;
  if (builder->relays.len < relay_count) return false;
  HeaplessVec_RouteHop_MAX_HOPS hops;
  heapless_vec_routehop_init(&hops);
  // Add relay hops
  for (size_t i = 0; i < relay_count; i++) 
  {
    size_t relay_index = i % builder->relays.len;
    if (!heapless_vec_routehop_push(&hops, &builder->relays.data[relay_index])) return false;
  }
  // Add destination hop
  if (!heapless_vec_routehop_push(&hops, destination)) return false;
  route_out->hops = hops;
  return true;
}

// RouteBuilder relay_count function
static inline size_t route_builder_relay_count(const RouteBuilder *builder) 
{
  return builder->relays.len;
}

// RouteBuilder default constructor
static inline void route_builder_default(RouteBuilder *builder) 
{
  route_builder_new(builder);
}
